### ISSUE TITLE
Does the QP module back-propagate gradient?

### ISSUE THREAD

**Issue by smallbox120** [icoz69/DeepEMD#13]
In file Network.py line 110:

           for i in range(num_query):
                     for j in range(num_proto):

                    _, flow = emd_inference_opencv(1 - similarity_map[i, j, :, :], weight_1[i, j, :], weight_2[j, i, :])

                    similarity_map[i, j, :, :] =(similarity_map[i, j, :, :])*torch.from_numpy(flow).cuda()

            temperature=(self.args.temperature/num_node)

            logitis = similarity_map.sum(-1).sum(-1) *  temperature

The calculation of the flow, i.e., the best match between features, is conducted when forward logits and losses, but does the back-propagation go through the 'emd_inference_opencv' module? If I understand correctly, this module is run on cpu() and numpy, therefore no gradient is tracked.

**Comment by icoz69** [comment]
correct. when you are using opencv, you omit the gradients through the solver. we have tips in the readme file.
